components/places/src/bookmark_sync/incoming.rs

```diff
@@ -521,83 +471,122 @@ impl<'a> IncomingApplicator<'a> {
-        val.as_i64()
-    } else if val.is_u64() {
-        Some(val.as_u64().unwrap() as i64)
-    } else if val.is_string() {
-        set_reupload(validity);
-        if let Ok(n) = val.as_str().unwrap().parse() {
-            Some(n)
-        } else {
-            None
+// Fixup incoming keywords by lowercasing them and removing surrounding whitespace
+//
+// Like Desktop, we don't reupload if a keyword needs to be fixed-up
+// trailing whitespace, or isn't lowercase.
+fn fixup_optional_keyword(val: &mut JsonValue, validity: &mut SyncedBookmarkValidity) {
+    match val {
+        JsonValue::String(s) => {
+            let fixed = s.trim().to_lowercase();
+            if fixed.is_empty() {
+                *val = JsonValue::Null;
+            } else if fixed != *s {
+                *val = JsonValue::String(fixed);
+            }
+        }
+        JsonValue::Null => (),
+        _ => {
+            set_reupload(validity);
+            *val = JsonValue::Null;
         }
-    } else if val.is_null() {
-        None
-    } else {
-        set_reupload(validity);
-        None
     }
 }
 
-fn unpack_optional_keyword(
-    key: &str,
-    data: &JsonValue,
-    validity: &mut SyncedBookmarkValidity,
-) -> Option<String> {
-    match &data[key] {
-        JsonValue::String(ref s) => {
-            // Like Desktop, we don't reupload if a keyword has leading or
-            // trailing whitespace, or isn't lowercase.
-            let k = s.trim();
-            if k.is_empty() {
-                None
-            } else {
-                Some(k.to_lowercase())
+fn fixup_optional_tags(val: &mut JsonValue, validity: &mut SyncedBookmarkValidity) {
+    match val {
+        JsonValue::Array(ref tags) => {
+            let mut valid_tags = HashSet::with_capacity(tags.len());
+            for v in tags {
+                if let JsonValue::String(tag) = v {
+                    let tag = match validate_tag(tag) {
+                        ValidatedTag::Invalid(t) => {
+                            log::trace!("Incoming bookmark has invalid tag: {:?}", t);
+                            set_reupload(validity);
+                            continue;
+                        }
+                        ValidatedTag::Normalized(t) => {
+                            set_reupload(validity);
+                            t
+                        }
+                        ValidatedTag::Original(t) => t,
+                    };
+                    if !valid_tags.insert(tag) {
+                        log::trace!("Incoming bookmark has duplicate tag: {:?}", tag);
+                        set_reupload(validity);
+                    }
+                } else {
+                    log::trace!("Incoming bookmark has unexpected tag: {:?}", v);
+                    set_reupload(validity);
+                }
+            }
+            if valid_tags.len() != tags.len() {
+                *val = JsonValue::Array(valid_tags.into_iter().map(JsonValue::from).collect());
```

Comment:
nitpick: Let's do this unconditionally; I think `valid_tags` might still have normalized tags even when `valid_tags.len() == tags.len()`?